### PR TITLE
Add CRD conversion webhook and move example

### PR DESCRIPTION
**Pull request by yaocw2020**
- Add CRD conversion webhook pkg
- Move example to harvester/webhook-sample

### CODE REVIEW

File: pkg/server/server.go [hunk 1]
```diff
@@ -171,7 +144,152 @@ func (s *AdmissionWebhookServer) listenAndServe(clients *clients.Clients, handle
...
+		}
+		if _, err := crdClient.Update(crdCopy); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// RegisterValidators registers validator to the webhook server.
```
Comment (w13915984028) [inline pkg/server/server.go:251]: we may add a flag to server, when `start()` set the flag to true, the following 3 registers check the flag and report error when server has been started.

File: pkg/server/server.go [hunk 2]
```diff
@@ -99,65 +110,27 @@ func (s *AdmissionWebhookServer) listenAndServe(clients *clients.Clients, handle
...
-					SideEffects:             &sideEffectClassNone,
-					AdmissionReviewVersions: []string{"v1", "v1beta1"},
-				},
-			},
+		s.caBundle = secret.Data[corev1.TLSCertKey]
+		// configure admitter webhook
+		validatingWebhookConfiguration := s.validatingWebhookConfiguration()
+		mutatingWebhookConfiguration := s.mutatingWebhookConfiguration()
+		if validatingWebhookConfiguration != nil || mutatingWebhookConfiguration != nil {
+			if err := apply.WithOwner(secret).ApplyObjects(validatingWebhookConfiguration, mutatingWebhookConfiguration); err != nil {
```
Comment (w13915984028) [inline pkg/server/server.go:123]: should `configureCRDConversionWebhook` related configuration also be applied to `secret` ?

File: pkg/server/server.go [hunk 3]
```diff
@@ -99,65 +110,27 @@ func (s *AdmissionWebhookServer) listenAndServe(clients *clients.Clients, handle
 		// Sleep here to make sure server is listening and all caches are primed
 		time.Sleep(15 * time.Second)
```
Comment (w13915984028) [inline pkg/server/server.go:117]: it is not related to this PR, but `sleep 15 to wait http server is ready` is tricky, check a flag or use a channel to sync will be a further enhancement

**Review by w13915984028 — APPROVED** [review]
General GOOD, leave few comments, thanks.

File: pkg/server/server.go (hunk 2 above)
Comment (yaocw2020) [reply pkg/server/server.go:123]: No, we configure CRD conversion webhook in the CRD.

File: pkg/server/server.go (hunk 3 above)
Comment (yaocw2020) [reply pkg/server/server.go:117]: Great consideration, I have filed an issue to track it.

File: README.md [hunk 4]
```diff
@@ -1,15 +1,6 @@
...
-The example implements a simple pod validator and pod mutator.
-- pod validator: log when a pod is deleted.
-- pod mutator: add label `example:example` when a pod is created.
-
-### How to deploy
-```shell
-export KUBECONFIG=<your kubeconfig>
-make apply
-```
+Go to [harvester/webhook-sample](https://github.com/harvester/webhook-sample) to look a simple example.
```
Comment (guangbochen) [inline README.md:6]: remove this or point to the https://github.com/harvester/network-controller-harvester/tree/master/pkg/webhook

File: go.mod [hunk 5]
```diff
@@ -9,22 +9,22 @@ replace (
...
 	k8s.io/kube-aggregator => k8s.io/kube-aggregator v0.23.7
 	k8s.io/kube-openapi => k8s.io/kube-openapi v0.0.0-20211115234752-e816edb12b65
-
 )
 
 require (
 	github.com/gorilla/mux v1.8.0
 	github.com/rancher/dynamiclistener v0.3.5
-	github.com/rancher/wrangler v1.0.1-0.20220520195731-8eeded9bae2a
+	github.com/rancher/wrangler v1.0.1-0.20230131212012-76adc44fca0c
```
Comment (guangbochen) [inline go.mod]: use the official tag v1.1.1 instead, btw can we bump to golang v1.19 if it is an easy change?

File: pkg/config/config.go [hunk 6]
```diff
@@ -1,6 +1,6 @@
 package config
 
-// Options for the admission webhook server
+// Options for the admitter webhook server
```
Comment (guangbochen) [inline pkg/config/config.go]: I think `admission` is a better word to be used in here and below, would prefer to update if it is easy to change.

File: pkg/server/server.go [hunk 7]
```diff
@@ -99,65 +115,27 @@ func (s *AdmissionWebhookServer) listenAndServe(clients *clients.Clients, handle
...
-						},
-						CABundle: secret.Data[corev1.TLSCertKey],
-					},
-					Rules:                   validationRules,
-					FailurePolicy:           &failPolicyFail,
-					SideEffects:             &sideEffectClassNone,
-					AdmissionReviewVersions: []string{"v1", "v1beta1"},
-				},
-			},
+		s.caBundle = secret.Data[corev1.TLSCertKey]
```
Comment (guangbochen) [inline pkg/server/server.go:118]: will the CA get expired and does it get renewed automatically?

**Review by futuretea — APPROVED** [review]
LGTM

**Review by starbops — APPROVED** [review]
LGTM, thanks!

File: README.md (hunk 4 above)
Comment (yaocw2020) [reply README.md:6]: I write the webhook-sample https://github.com/yaocw2020/webhook-sample/tree/master and I'll migrate to harvester team.

File: go.mod (hunk 5 above)
Comment (yaocw2020) [reply go.mod:17]: The secret is managed by github.com/rancher/dynamiclistener and we need dive deeply to learn how it works.

File: pkg/server/server.go (hunk 7 above)
Comment (yaocw2020) [reply pkg/server/server.go:118]: The secret is managed by github.com/rancher/dynamiclistener and we need to dive deeply to learn how it works.

File: README.md (hunk 4 above)
Comment (yaocw2020) [reply README.md:6]: After the PR merged, I'll update the dependency in the sample and migrate it to harvester team.

**Review by guangbochen — APPROVED** [review]

**Pull request merged**